vortex-alp/src/array.rs

```diff
@@ -43,120 +35,88 @@ impl ALPArray {
+        } else {
+            vortex_bail!("ALP can only encode primitive arrays");
         }
     }
 
-    pub fn encoded(&self) -> &ArrayRef {
-        &self.encoded
+    pub fn encoded(&self) -> Array {
+        self.array()
+            .child(0, &self.metadata().dtype)
+            .expect("Missing encoded array")
     }
 
     pub fn exponents(&self) -> &Exponents {
-        &self.exponents
+        &self.metadata().exponents
     }
 
-    pub fn patches(&self) -> Option<&ArrayRef> {
-        self.patches.as_ref()
+    pub fn patches(&self) -> Option<Array> {
+        self.metadata().has_patches.then(|| {
+            self.array()
+                .child(1, self.dtype())
+                .expect("Missing patches with present metadata flag")
+        })
     }
 }
 
-impl Array for ALPArray {
-    impl_array!();
-
-    #[inline]
-    fn len(&self) -> usize {
-        self.encoded.len()
-    }
-
-    #[inline]
-    fn is_empty(&self) -> bool {
-        self.encoded.is_empty()
-    }
-
-    #[inline]
-    fn dtype(&self) -> &DType {
-        &self.dtype
-    }
-
-    #[inline]
-    fn stats(&self) -> Stats {
-        Stats::new(&self.stats, self)
-    }
-
-    #[inline]
-    fn with_compute_mut(
-        &self,
-        f: &mut dyn FnMut(&dyn ArrayCompute) -> VortexResult<()>,
-    ) -> VortexResult<()> {
-        f(self)
-    }
-
-    #[inline]
-    fn encoding(&self) -> EncodingRef {
-        &ALPEncoding
-    }
-
-    #[inline]
-    fn nbytes(&self) -> usize {
-        self.encoded().nbytes() + self.patches().map(|p| p.nbytes()).unwrap_or(0)
-    }
-
-    fn serde(&self) -> Option<&dyn ArraySerde> {
-        Some(self)
+impl ArrayValidity for ALPArray<'_> {
+    fn is_valid(&self, index: usize) -> bool {
+        self.encoded().with_dyn(|a| a.is_valid(index))
     }
 
-    fn walk(&self, walker: &mut dyn ArrayWalker) -> VortexResult<()> {
-        walker.visit_child(self.encoded())
+    fn logical_validity(&self) -> LogicalValidity {
+        self.encoded().with_dyn(|a| a.logical_validity())
     }
 }
 
-impl ArrayDisplay for ALPArray {
-    fn fmt(&self, f: &mut ArrayFormatter) -> std::fmt::Result {
-        f.property("exponents", format!("{:?}", self.exponents()))?;
-        f.child("encoded", self.encoded())?;
-        f.maybe_child("patches", self.patches())
+impl ArrayFlatten for ALPArray<'_> {
+    fn flatten<'a>(self) -> VortexResult<Flattened<'a>>
+    where
+        Self: 'a,
+    {
+        decompress(self).map(Flattened::Primitive)
     }
 }
 
-impl ArrayValidity for ALPArray {
-    fn logical_validity(&self) -> Validity {
-        self.encoded().logical_validity()
-    }
-
-    fn is_valid(&self, index: usize) -> bool {
-        self.encoded().is_valid(index)
+impl AcceptArrayVisitor for ALPArray<'_> {
+    fn accept(&self, visitor: &mut dyn ArrayVisitor) -> VortexResult<()> {
+        visitor.visit_child("encoded", &self.encoded())?;
+        if self.metadata().has_patches {
```

Comment:
At this point you should just rely on the `self.patches()` since you already run the check inside there